src/core/producerconsumerqueue.h

```diff
@@ -0,0 +1,156 @@
+{
+public:
+    /**
+     * Constructs a ProducerConsumerQueue()
+     * @param[in] capacity the maximum size of this queue
+     */
+    ProducerConsumerQueue(int capacity)
+        : m_capacity(capacity)
+    {
+        assert(m_capacity > 0);
+    };
+
+    /**
+     * Constructs an empty ProducerConsumerQueue with capacity 0
+     * In nonblocking mode all operations will immediately fail
+     * In blocking mode all operations will fail an assertion to avoid blocking forever
+     */
+    ProducerConsumerQueue()
+        : m_capacity(0){};
+    ~ProducerConsumerQueue(){};
+
+    /**
+     * Push an element to the back of the queue
+     * Blocking producer mode: will always eventually succeed
+     * Non-blocking producer mode: Push() returns failure when the queue is at capacity
+     * @param[in] data the data to be pushed
+     * @return the success of the operation
+     * @see WorkerMode
+     */
+    template <typename TT>
+    bool Push(TT&& data, WorkerMode mode = m_producer_mode)
+    {
+        // TT needed for perfect forwarding to vector::push_back
+
+        // attempting a push to a queue of capacity 0 is likely unintended
+        assert(m_capacity > 0);
+
+        {
+            std::unique_lock<std::mutex> l(m_queue_lock);
+            if (m_data.size() >= m_capacity) {
+                if (mode == WorkerMode::NONBLOCKING) {
+                    return false;
+                }
+
+                m_producer_cv.wait(l, [&]() { return m_data.size() < m_capacity; });
+            }
+
+            m_data.push_back(std::forward<TT>(data));
+        }
+        m_consumer_cv.notify_one();
+        return true;
+    };
+
+    /**
+     * Try to pop the oldest element from the front of the queue, if present
+     * Blocking consumer mode: will always eventually succeed
+     * Nonblocking consumer mode: Pop() returns failure when the queue is empty
+     * @param[out] the data popped, if the operation was successful
+     * @return the success of the operation
+     * @see WorkerMode
+     */
+    bool Pop(T& data, WorkerMode mode = m_consumer_mode)
+    {
+        // attempting a pop from a queue of capacity 0 is likely unintended
+        assert(m_capacity > 0);
+
+        {
+            std::unique_lock<std::mutex> l(m_queue_lock);
+            if (m_data.size() <= 0) {
+                if (mode == WorkerMode::NONBLOCKING) {
+                    return false;
+                }
+
+                m_consumer_cv.wait(l, [&]() { return m_data.size() > 0; });
+            }
+
+            data = std::move(m_data.front());
+            m_data.pop_front();
+        }
+        m_producer_cv.notify_one();
+        return true;
+    }
+
+    /**
+     * Shortcut for bool Pop(T&) when consumer mode is blocking
+     * This must always succeed and thus may only be called in producer blocking mode
+     * @return the element popped
+     */
+    T Pop()
+    {
+        static_assert(m_consumer_mode == WorkerMode::BLOCKING, "");
```

Comment:
In commit "Implement a thread-safe FIFO (producer/consumer style) queue" (d8106146e7511837decb4b45de35dec654454b69)

Could return std::future\<T> in the case of a non-blocking Pop() to support it instead of having this asymmetry.

Reply:
Doing that would complicate internal implementation a bit - you'd have to hang on to the promise that satisfies the future internally - and either you'd have too many of those and need to block - or potentially allow your buffer holding promises to grow unbounded - so actually it might not be safely implementable. 

Also I don't really see an immediate use case for this - you'd have to later wait on the future (blocking or non-blocking) - but you could just alternately wait on the queue. I can't think of any use cases right now where you'd want to reserve a place in line in a non-blocking fashion and then later claim that item.